Add `set_code` kernel procedure

frisitano opened the pull request:
This PR introduces the `set_code` procedure.  We also introduce a `type` procedure which returns the type of the account.  We should consider if we want to expose this publicly or keep it private?

bobbinth commented on miden-lib/asm/sat/layout.masm:
```diff
@@ -389,6 +389,16 @@ export.set_acct_nonce
+#! Output: []
+#!
+#! - CODE_ROOT is the code root to be set.
+export.set_acct_code_root
+    push.ACCT_CODE_ROOT_PTR mem_storew dropw
+end
```
This would update the account code root immediately, right? If so, I wonder if this is the right approach. The reason is that for access checks (#79) we'd need to check if a procedure exists against the code root. So, changing it immediately would mean that the user would be able to do a "hot swap" of the code, including any tx authentication procedures. This may or may not be a problem - but I wonder if we should be conservative here.

So, I guess we need to decide:
1. Do we update the code root immediately?
2. Do we delay the update until the epilogue? (i.e., cache the new code root somewhere, and copy it over in the epilogue).

My preliminary thinking is that option 2 is better as it is simpler to think through and much less complex (even from the advice provider standpoint).

frisitano replied:
I think either approach is valid and (most likely) practical. I recognise that there may be some additional complexity around advice provider interaction with immediate code root update. I don't have a strong preference either way.  I'll push on with option 2 for now.  I think the decision should be made based on UX.

 One question on this topic is whether we allow the `set_code` procedure to be called multiple times in a transaction and what the expected behaviour should be in this scenario?

bobbinth commented on miden-lib/asm/sat/account.masm:
```diff
@@ -66,3 +84,46 @@ end
+    # => [acct_id]
+
+    # compute the account type
+    u32split swap drop push.30 u32checked_shr
```
We should use `u32checked_shr.30` here it is way more efficient than the one which reads the number of bits to shift by from the stack (i.e., 4 vs. 47 cycles).

bobbinth reviewed (changes requested):
Thank you! Looks good! I left a couple of comments inline.

Regarding exporting of `type`. I'm not sure it makes sense as we don't export the enum constants. We could export dedicated methods  - e.g., `is_upgradable_account`, `is_fungible_asset_faucet` etc.

frisitano commented:
I've updated the PR such that `set_code` works as follows:

- I've introduced a `ACCT_NEW_CODE_ROOT_PTR` memory entry.
- In the prologue we set `ACCT_NEW_CODE_ROOT_PTR` to the initial account code roo.
- If `account::set_code` is invoked then `ACCT_NEW_CODE_ROOT_PTR` is updated accordingly.
- In the epilogue we read `ACCT_NEW_CODE_ROOT_PTR` and save it in `ACCT_CODE_ROOT_PTR`.
- If `account::set_code` has not been invoked this will result in a no-op.
- No conditional statements required in this pattern.

bobbinth commented on miden-lib/asm/sat/epilogue.masm:
```diff
@@ -149,6 +149,26 @@ proc.process_created_notes
+
+    # set the account code root to the new account code root (may not have changed)
+    exec.layout::set_acct_code_root
+    # => [NEW_ACCT_CODE_ROOT]
```
Should this be `# => []`?

frisitano replied:
yes - updated.

bobbinth reviewed (approved):
Looks good! Thank you! I left one non-blocking comment inline.

> One question on this topic is whether we allow the `set_code` procedure to be called multiple times in a transaction and what the expected behaviour should be in this scenario?

I'm thinking it is OK to call it multiple time with the last call taking precedence. The alternative could be to allow only one call and fail on subsequent calls. If an argument can be made that it is preferred - I'm fine with this approach too.

Also, what did you think about my comment in https://github.com/0xPolygonMiden/miden-base/pull/91#pullrequestreview-1385924172?
